sync15-adapter/src/token.rs

```diff
@@ -0,0 +1,133 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use hawk;
+
+use reqwest::{Client, Request, Url};
+use hyper::header::{Authorization, Bearer};
+use error::{self, Result};
+use std::fmt;
+use std::borrow::{Borrow, Cow};
+use util::ServerTimestamp;
+
+/// Tokenserver's timestamp is X-Timestamp and not X-Weave-Timestamp.
+header! { (RetryAfter, "Retry-After") => [f64] }
+
+/// Tokenserver's timestamp is X-Timestamp and not X-Weave-Timestamp. The value is in seconds.
+header! { (XTimestamp, "X-Timestamp") => [ServerTimestamp] }
+
+/// OAuth tokenserver api uses this instead of X-Client-State.
+header! { (XKeyID, "X-KeyID") => [String] }
+
+#[derive(Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct TokenserverToken {
+    pub id: String,
+    pub key: String,
+    pub api_endpoint: String,
+    pub uid: u64,
+    pub duration: u64,
+    // This is treated as optional by at least the desktop client,
+    // but AFAICT it's always present.
+    pub hashed_fxa_uid: String,
+}
+
+/// This is really more of a TokenAuthenticator.
+pub struct TokenserverClient {
+    token: TokenserverToken,
+    server_timestamp: ServerTimestamp,
+    credentials: hawk::Credentials,
+}
+
+// hawk::Credentials doesn't implement debug -_-
+impl fmt::Debug for TokenserverClient {
+    fn fmt(&self, f: &mut fmt::Formatter) -> ::std::result::Result<(), fmt::Error> {
+        f.debug_struct("TokenserverClient")
+         .field("token", &self.token)
+         .field("server_timestamp", &self.server_timestamp)
+         .field("credentials", &"(omitted)")
+         .finish()
+    }
+}
+
+fn token_url(base_url: &str) -> Result<Url> {
+    let mut url = Url::parse(base_url)?;
+    // kind of gross but avoids problems if base_url has a trailing slash.
+    url.path_segments_mut()
+        // We can't do anything anyway if this is the case.
+       .map_err(|_| error::unexpected("Bad tokenserver url (cannot be base)"))?
+       .extend(&["1.0", "sync", "1.5"]);
+    Ok(url)
+}
+
+impl TokenserverClient {
+    #[inline]
+    pub fn server_timestamp(&self) -> ServerTimestamp {
+        self.server_timestamp
+    }
+
+    #[inline]
+    pub fn token(&self) -> &TokenserverToken {
+        &self.token
+    }
+
+    pub fn new(request_client: &Client, base_url: &str, access_token: String, key_id: String) -> Result<TokenserverClient> {
```

Comment:
Playing a bit more with this, I'm not sure I'm a fan of `new` making an HTTP request when you construct it...what happens if the network is flaky, or the server is down? Could this be lazy, so that we don't fail to create the objects we need for a sync?

Reply:
I tweaked this in #42 - let me know what you think!